From 5e199fe39f7bf75a64f094b2e0a8b00c4977ff4d Mon Sep 17 00:00:00 2001
From: Ruoyu Ying <ruoyu.ying@intel.com>
Date: Thu, 29 Jun 2023 13:28:05 -0400
Subject: [PATCH 04/16] fix: fix sha index issue in IMA

* Add extra check for sha1 index to avoid using negative number as
index

* Add extra check to tdx device node to skip IMA RTMR for efi guest

Signed-off-by: Ruoyu Ying <ruoyu.ying@intel.com>
---
 security/integrity/ima/ima_crypto.c |  2 +-
 security/integrity/ima/ima_init.c   | 14 +++++++++-----
 2 files changed, 10 insertions(+), 6 deletions(-)

diff --git a/security/integrity/ima/ima_crypto.c b/security/integrity/ima/ima_crypto.c
index b62ab0c6a8bb..665c5b921d5a 100644
--- a/security/integrity/ima/ima_crypto.c
+++ b/security/integrity/ima/ima_crypto.c
@@ -181,7 +181,7 @@ int __init ima_init_crypto(void)
 		}
 	}
 
-	if (ima_sha1_idx >= NR_BANKS(ima_tpm_chip)) {
+	if (ima_sha1_idx >= 0 && ima_sha1_idx >= NR_BANKS(ima_tpm_chip)) {
 		if (ima_hash_algo == HASH_ALGO_SHA1) {
 			ima_algo_array[ima_sha1_idx].tfm = ima_shash_tfm;
 		} else {
diff --git a/security/integrity/ima/ima_init.c b/security/integrity/ima/ima_init.c
index c9b6c119b8f0..7294ad076b07 100644
--- a/security/integrity/ima/ima_init.c
+++ b/security/integrity/ima/ima_init.c
@@ -128,12 +128,16 @@ int __init ima_init(void)
 	ima_tpm_chip = tpm_default_chip();
 	if (!ima_tpm_chip) {
 #ifdef CONFIG_INTEL_TDX_GUEST
-		pr_info("No TPM chip found, Checking TDX instead!\n");
-		/* Only support RTMR case while setting ima_hash=sha384
-		 * in kernel cmdline
-		 */
-		if (ima_hash_algo == HASH_ALGO_SHA384)
+		u32 eax, sig[3];
+		cpuid_count(TDX_CPUID_LEAF_ID, 0, &eax, &sig[0], &sig[2],  &sig[1]);
+
+		if (!memcmp(TDX_IDENT, sig, sizeof(sig)) && ima_hash_algo == HASH_ALGO_SHA384) {
+		    /* Only support RTMR case while setting ima_hash=sha384
+		    * in kernel cmdline
+		    */
+			pr_info("No TPM chip found, Checking TDX instead!\n");
 			ima_tpm_chip = tdx_rtmr_device();
+		}
 #endif
 		if (!ima_tpm_chip) {
 		        pr_info("No TPM chip found, activating TPM-bypass!\n");
-- 
2.40.1

